docs/spec/lightclient/verification.md

```diff
@@ -844,9 +915,130 @@ If on the blockchain the validator set of the block at height
 
 
 
+## Supporting the IBC Relayer
 
+We might need to download and verify headers whose heights are smaller
+than lightStore.LatestVerified.Height. To formalize
+the solution, we start with the following methods for the LightStore
 
 
+```go
+func (ls LightStore) LatestPrevious(height Height) (LightBlock, bool) 
+```
+- Expected postcondition
+  - returns a light block *lb* that satisfies:
+      - *lb* is in lightStore
+      - *lb* is verified and not expired
+	  - *lb.Header.Height < height*
+	  - for all *b* in lightStore s.t. *b* is verified and not expired it
+        holds *lb.Header.Height >= b.Header.Height*
+   - or false in the second argument if
+    the LightStore does not contain such an *lb*.
+ 
+
+```go
+func (ls LightStore) MinVerified() (LightBlock, bool) 
+```
+- Expected postcondition
+  - returns a light block *lb* that satisfies:
+      - *lb* is in lightStore
+      - *lb* is verified
+	  - *lb.Header.Height* is minimal in the lightStore
+	  - **TODO:** according to this it might be expired (outside the
+        trusting period). I believe this is safe. Are there reasons we
+        should not do that?
+   - or false in the second argument if
+    the LightStore does not contain such an *lb*.
+ 
+If a height that is smaller than the smallest height in the lightstore
+is required, we check the hashes backwards. This is done with the
+following function:
+ 
+#### **[LCV-FUNC-BACKWARDS.1]**:
+```go
+func Backwards(primary PeerID, lightStore LightStore,
+                    targetHeight Height) (LightStore, Result) {
+  
+  lb,res = lightStore.MinVerified()
+  if res = false {
+    return (lightStore, ResultFailure)
+  }
+
+  latest := lb.Header
+  for i := lb.Header.height - 1; i >= targetHeight; i-- {
+    // here we download height-by-height. We might first download all
+	// headers down to targetHeight and then check them.
+    current := FetchLightBlock(primary,i)
+    if (hash(current) != latest.Header.LastBlockId) {
+      return (lightStore, ResultFailure)
+    }
+	else {
+	  lightStore.Update(current, StateVerified)
+	}
+    latest := current
+  }
+  return (lightStore, ResultSuccess)
+ }
+```
+
+The following function just decided based on the required height which
+method should be used.
+
+#### **[LCV-FUNC-IBCMAIN.1]**:
+```go
+func Main(primary PeerID, lightStore LightStore,
+                    targetHeight Height) (LightStore, Result) {
+  if targetHeight > lightStore.LatestVerified.height {
+    return VerifyToTarget(primary, lightStore, targetHeight)
+  }
+  else if targetHeight = lightStore.LatestVerified.height {
```

Comment:
This should probably be just check if the targetHeight is already in the store.

Reply:
I think the main decision tree should be as follows: 1) is `targetHeight` already in the store; if this is the case return the block, else 2) if `targetHeight >  lightStore.LatestVerified.height` do normal forward verification business, else 3) if  `targetHeight > some trusted height minH from the store` then do forward verification starting from `minH`, else 4) do the backward verification from some trusted header mH from the store, but being careful that backward verification must terminate within trusted period of mH. So picking mH as minimum trusted header might not be the best idea. In practice, I would guess that most of the verification requests would be addressed by scenarios 1-3 (at least for relayer cases).

Reply:
OK.  
A clarification for point 1). `targetHeight` needs to be in store **and** must be verified (I adapted to code accordingly). It might be that we previously downloaded it but could not verify it, and then did some bisection, and eventually jumped over it. Apalache found such traces, and they are fine.  
Regarding point 4, I am not sure. If any header is set to trusted, we treat it as if it comes from the blockchain. By assumption on hashes, all previous blocks that pass the tests, necessarily are also from the chain. The problem of course is whether the application should the use the header of targetHeight, when it was expired. I think this is a question for the specific application. It might be application-specific whether the information in such a block is still usable.